base_exception/models/base_exception.py

```diff
@@ -74,14 +52,132 @@ class ExceptionRule(models.Model):
+        """Filter exception.rules.
+
+        By default, only the rules with the correct model
+        will be used.
+        """
+        return [('model', '=', self._name)]
+
+    @api.multi
+    def detect_exceptions(self):
+        """List all exception_ids applied on self
+        Exception ids are also written on records
+        If self is empty, check exceptions on all records.
+        """
+        rules = self.env['exception.rule'].sudo().search(
+            self._rule_domain())
+        all_exception_ids = []
+        for rule in rules:
+            records_with_exception = self._detect_exceptions(rule)
+            reverse_field = self._reverse_field()
+            if self:
+                commons = self and rule[reverse_field]
+                to_remove = commons - records_with_exception
+                to_add = records_with_exception - commons
+                to_remove_list = [(3, x.id, _) for x in to_remove]
+                to_add_list = [(4, x.id, _) for x in to_add]
+                rule.write({reverse_field: to_remove_list + to_add_list})
+            else:
+                rule.write({
+                    reverse_field: [(6, 0, records_with_exception.ids)]
+                })
+            if records_with_exception:
+                all_exception_ids.append(rule.id)
+        return all_exception_ids
+
+    @api.model
+    def _exception_rule_eval_context(self, rec):
+        return {'self': self.pool.get(rec._name),
+                'object': rec,
+                'obj': rec,
+                'pool': self.pool,
+                'env': self.env,
+                'cr': self.env.cr,
+                'uid': self.env.uid,
+                'user': self.env.user,
+                'time': time,
+                # copy context to prevent side-effects of eval
+                'context': self.env.context.copy()}
+
+    @api.model
+    def _rule_eval(self, rule, rec):
+        expr = rule.code
+        space = self._exception_rule_eval_context(rec)
+        try:
+            safe_eval(expr,
+                      space,
+                      mode='exec',
+                      nocopy=True)  # nocopy allows to return 'result'
+        except Exception, e:
+            raise UserError(
+                _('Error when evaluating the exception.rule '
+                  'rule:\n %s \n(%s)') % (rule.name, e))
+        return space.get('failed', False)
+
+    @api.multi
+    def _detect_exceptions(self, rule):
+        if rule.exception_type == 'by_py_code':
+            return self._detect_exceptions_by_py_code(rule)
+        elif rule.exception_type == 'by_domain':
+            return self._detect_exceptions_by_domain(rule)
+
+    @api.multi
+    def _detect_exceptions_by_py_code(self, rule):
+        """
+            Find exceptions found on self.
+            If self is empty, check on all records.
+        """
+        domain = [('ignore_exception', '=', False)]
+        if self:
+            domain.append(('id', 'in', self.ids))
```

Comment:
@florian-dacosta You can maybe avoid duplicate code for the domain of the search by adding a method to get the base domain